### PR TITLE
Enforce schema properties order in JsonBuilder so Pegasus schema and Avro schema properties would have deterministic order

### PR DESCRIPTION
**Pull request by junchuanwang**
Currently when translating to Avro or Pegasus schema, nested properties wouldn't have deterministic and based on its implementation (DataMap type), the order is given from HashMap, which is indeterministic.

This change would affect [pegasus to avro] translation AND [avro to peagsus] translation when outputting string (only for properties encoding). It wouldn't affect memory representation.

wc-test:
https://crt.prod.linkedin.com/#/testing/executions/7037b41c-8501-4db9-9237-1f54c5964c68/execution

### CODE REVIEW

File: data/src/main/java/com/linkedin/data/codec/AbstractJacksonDataCodec.java [hunk 1]
```diff
@@ -296,12 +296,20 @@ public void startMap(DataMap map) throws IOException
       _generator.writeStartObject();
     }
 
+    @Override
+    public Iterable<Map.Entry<String, Object>> orderMap(DataMap map)
```
Comment (nickibi) [inline data/src/main/java/com/linkedin/data/codec/AbstractJacksonDataCodec.java]: I am curious about how this method is being called?

File: data-avro/src/test/java/com/linkedin/data/avro/TestSchemaTranslator.java [hunk 2]
```diff
@@ -1660,6 +1660,46 @@ public void testToAvroSchema_testTypeRefAnnotationPropagation(String schemaBefor
...
+          {
+              " { " +
+              "   \"type\" : \"record\", " +
+              "   \"name\" : \"Foo\", " +
+              "   \"fields\" : [ { " +
+              "     \"name\" : \"field1\", " +
+              "     \"type\" : \"int\", " +
+              "     \"b_customerAnnotation\" : \"f1\", " +
+              "     \"c_customerAnnotation\" : \"f1\", " +
+              "     \"a_customerAnnotation\" : \"f1\" " +
```
Comment (evanw555) [inline data-avro/src/test/java/com/linkedin/data/avro/TestSchemaTranslator.java]: nit: I think you mean to say `x_customAnnotation`?

File: data/src/main/java/com/linkedin/data/schema/JsonBuilder.java [hunk 3]
```diff
@@ -282,7 +280,10 @@ public void writeProperties(Map<String, ?> value) throws IOException
   {
     if (value.isEmpty() == false)
     {
-      for (Map.Entry<String, ?> entry : value.entrySet())
+      List<Map.Entry<String, ?>> orderedProperties =  value.entrySet().stream()
```
Comment (evanw555) [inline data/src/main/java/com/linkedin/data/schema/JsonBuilder.java]: The unit tests seems to imply this, however I'd like to confirm; this method is only for writing custom properties, right? Since this is a generic `JsonBuilder`, does it have any sense of what "properties" are? Or is this just for writing any generic subtree of the JSON object?

File: data/src/main/java/com/linkedin/data/codec/AbstractJacksonDataCodec.java [hunk 4]
```diff
@@ -296,6 +296,12 @@ public void startMap(DataMap map) throws IOException
...
```
Comment (karthikbalasub) [inline data/src/main/java/com/linkedin/data/codec/AbstractJacksonDataCodec.java]: This is not a good idea. The jackson codec is used for serializing data on the servers and making this change here will affect performance of services.

File: data/src/main/java/com/linkedin/data/schema/JsonBuilder.java (hunk 3 above)
Comment (junchuanwang) [reply data/src/main/java/com/linkedin/data/schema/JsonBuilder.java]: this is just for "writing  generic subtree of the JSON object", any concerns?

File: data-avro/src/test/java/com/linkedin/data/avro/TestSchemaTranslator.java (hunk 2 above)
Comment (junchuanwang) [reply data-avro/src/test/java/com/linkedin/data/avro/TestSchemaTranslator.java]: yes sir.

File: data/src/main/java/com/linkedin/data/codec/AbstractJacksonDataCodec.java (hunk 4 above)
Comment (junchuanwang) [reply data/src/main/java/com/linkedin/data/codec/AbstractJacksonDataCodec.java]: Update code ,can you please check? Now I limit the scope of the change so it only applies when writing properties.

File: data/src/main/java/com/linkedin/data/codec/AbstractJacksonDataCodec.java [hunk 5]
```diff
@@ -213,12 +224,35 @@ public DataList readList(InputStream in) throws IOException
     return parse(_factory.createParser(in), mesg, locationMap);
   }
 
-  public void objectToJsonGenerator(Object object, JsonGenerator generator) throws IOException
```
Comment (karthikbalasub) [inline data/src/main/java/com/linkedin/data/codec/AbstractJacksonDataCodec.java]: Do not change public APIs. Keep this one and add the second API

File: data/src/main/java/com/linkedin/data/codec/AbstractJacksonDataCodec.java [hunk 6]
```diff
@@ -133,7 +133,18 @@ protected void writeObject(Object object, JsonGenerator generator) throws IOExce
 
   protected Data.TraverseCallback createTraverseCallback(JsonGenerator generator)
   {
-    return new JacksonTraverseCallback(generator);
+    return createTraverseCallback(generator, false);
+  }
+
+  /**
+   * Create {@link JacksonTraverseCallback} instance for Data object traverse
```
Comment (karthikbalasub) [inline data/src/main/java/com/linkedin/data/codec/AbstractJacksonDataCodec.java]: JacksonTraverseCallback -> Traverse.Callback
Avoid implementation details in the java doc

File: data/src/main/java/com/linkedin/data/codec/AbstractJacksonDataCodec.java [hunk 7]
```diff
@@ -213,12 +224,35 @@ public DataList readList(InputStream in) throws IOException
     return parse(_factory.createParser(in), mesg, locationMap);
   }
 
-  public void objectToJsonGenerator(Object object, JsonGenerator generator) throws IOException
+  /**
+   * Convert an object to Json format representation.
+   * @param object the object that needs to be converted
+   * @param generator the generator that could generate Json output based on the object
+   * @param mapKeyOrdered if the object contains map element (which could be nested map), can set this flag to order the
```
Comment (karthikbalasub) [inline data/src/main/java/com/linkedin/data/codec/AbstractJacksonDataCodec.java]: If true, then map elements in the object (can be the object itself or its nested elements) will be sorted by Key in the generated Json.

File: data/src/main/java/com/linkedin/data/codec/AbstractJacksonDataCodec.java [hunk 8]
```diff
@@ -213,12 +224,35 @@ public DataList readList(InputStream in) throws IOException
...
+   * @throws IOException
+   */
+  public void objectToJsonGenerator(Object object, JsonGenerator generator, boolean mapKeyOrdered) throws IOException
   {
-    Data.TraverseCallback callback = createTraverseCallback(generator);
+    Data.TraverseCallback callback = createTraverseCallback(generator, mapKeyOrdered);
     Data.traverse(object, callback);
   }
 
+  public static class JacksonTraverseCallbackMapKeyOrdered extends JacksonTraverseCallback
```
Comment (karthikbalasub) [inline data/src/main/java/com/linkedin/data/codec/AbstractJacksonDataCodec.java]: Why do you need this class? Can't you add a flag in JacksonTraverseCallback and a second constructor to set it?

File: data/src/main/java/com/linkedin/data/schema/JsonBuilder.java [hunk 9]
```diff
@@ -269,11 +267,24 @@ public void writeMapField(String fieldName, Map<String, ?> value, boolean requir
    */
   public void writeData(Object object) throws IOException
   {
-    _jacksonDataCodec.objectToJsonGenerator(object, _jsonGenerator);
+    _jacksonDataCodec.objectToJsonGenerator(object, _jsonGenerator, false);
   }
 
+  /**
+   * Write Data object. But if the Data Object contains DataMap, the result would output the map entries using sorted key order.
```
Comment (karthikbalasub) [inline data/src/main/java/com/linkedin/data/schema/JsonBuilder.java]: document that this applies to object also.

File: data/src/main/java/com/linkedin/data/codec/AbstractJacksonDataCodec.java (hunk 5 above)
Comment (junchuanwang) [reply data/src/main/java/com/linkedin/data/codec/AbstractJacksonDataCodec.java]: Good catch. Fail to notice it was public

File: data/src/main/java/com/linkedin/data/codec/AbstractJacksonDataCodec.java (hunk 6 above)
Comment (junchuanwang) [reply data/src/main/java/com/linkedin/data/codec/AbstractJacksonDataCodec.java]: I wouldn't say I totally agree with you in this one. But I made the change according to your suggestion/

The reason is that this method take in "generator" as input parameter in order to return Data.TraverseCallback, this is already very implementation specific and very detailed implementation description. Something unavoidable.

File: data/src/main/java/com/linkedin/data/codec/AbstractJacksonDataCodec.java (hunk 8 above)
Comment (junchuanwang) [reply data/src/main/java/com/linkedin/data/codec/AbstractJacksonDataCodec.java]: Because I need to override the 
`orderMap` method in the new class. 

I don't think I got you. How do I set the override method in the constructor?

Comment (junchuanwang) [reply data/src/main/java/com/linkedin/data/codec/AbstractJacksonDataCodec.java]: after off line discussion, adopted the suggestion

File: data/src/main/java/com/linkedin/data/schema/JsonBuilder.java (hunk 3 above)
Comment (evanw555) [reply data/src/main/java/com/linkedin/data/schema/JsonBuilder.java]: Would this affect the order in which all schema attributes are written, even non-custom properties? e.g. [namespace, name, type, fields] vs [type, fields, name, namespace]. Or are those written individually in the schema encoder? I'm concerned that this would affect the ordering of things that aren't custom properties

File: data/src/main/java/com/linkedin/data/template/JacksonDataTemplateCodec.java [hunk 10]
```diff
@@ -87,7 +87,7 @@ protected void dataTemplateToJsonGenerator(Object data,
     }
     else
     {
-      objectToJsonGenerator(data, generator);
+      objectToJsonGenerator(data, generator, false);
```
Comment (karthikbalasub) [inline data/src/main/java/com/linkedin/data/template/JacksonDataTemplateCodec.java]: You can revert this change now.

File: data/src/main/java/com/linkedin/data/schema/JsonBuilder.java (hunk 3 above)
Comment (junchuanwang) [reply data/src/main/java/com/linkedin/data/schema/JsonBuilder.java]: it shouldn't. non-customer properties are written using other code. This code only manage the properties, so all properties written as a trunk. shouldn't affect others

File: data/src/main/java/com/linkedin/data/template/JacksonDataTemplateCodec.java (hunk 10 above)
Comment (nickibi) [reply data/src/main/java/com/linkedin/data/template/JacksonDataTemplateCodec.java]: No need for this change.
